File: hw1/1/tf_05_test.py
import unittest
# from tf_05 import filter_chars_and_normalize
import tf_05

class TestTF05(unittest.TestCase):

    def setUp(self):
        tf_05.data = []
        tf_05.words = []
        tf_05.word_freqs = []

    def test_filter_chars_and_normalize(self):
        sentence = "This is a test. This test is only a test."
        expected = "this is a test  this test is only a test "
        tf_05.data = list(sentence)
        tf_05.filter_chars_and_normalize()
        self.assertEqual(expected, ''.join(tf_05.data))

    def test_scan(self):
        sentence = "this is a test  this test is only a test "
        expected = ['this', 'is', 'a', 'test', 'this', 'test', 'is', 'only', 'a', 'test']
        tf_05.data = list(sentence)
        tf_05.scan()
        self.assertEqual(expected, tf_05.words)

    def test_remove_stop_words(self):
        tf_05.words = ['this', 'is', 'a', 'test', 'this', 'test', 'is', 'only', 'a', 'test']
        tf_05.remove_stop_words()
        self.assertEqual(['test', 'test', 'test'], tf_05.words)

    def test_frequencies(self):
        tf_05.words = ['test', 'test', 'test']
        tf_05.frequencies()
        self.assertEqual([['test', 3]], tf_05.word_freqs)

if __name__ == '__main__':
    unittest.main()

File: hw1/3/tf_05_test.go
package main

import (
	"testing"
)

func TestFilterCharsAndNormalize(t *testing.T) {
	data = []rune("Hello, World! 123")
	FilterCharsAndNormalize()
	if string(data) != "hello  world  123" {
		t.Errorf("Expected 'hello  world  123', got '%s'", string(data))
	}
}

func TestScan(t *testing.T) {
	data = []rune("hello world 123")
	scan()
	expected := []string{"hello", "world", "123"}
	for i, word := range words {
		if word != expected[i] {
			t.Errorf("Expected '%s', got '%s'", expected[i], word)
		}
	}
}

func TestFrequencies(t *testing.T) {
	words = []string{"hello", "world", "hello"}
	frequencies()
	expected := map[string]int{"hello": 2, "world": 1}
	for word, count := range expected {
		if freqs[word] != count {
			t.Errorf("Expected count %d for word '%s', got %d", count, word, freqs[word])
		}
	}
}

func TestSortFrequencies(t *testing.T) {
	freqs = map[string]int{"hello": 2, "world": 1}
	sortFrequencies()
	if sortedFreqs[0].Word != "hello" || sortedFreqs[0].Count != 2 {
		t.Errorf("Expected 'hello' with count 2, got '%s' with count %d", sortedFreqs[0].Word, sortedFreqs[0].Count)
	}
}
